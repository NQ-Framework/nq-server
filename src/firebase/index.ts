export * from './firebase-auth-middleware';
