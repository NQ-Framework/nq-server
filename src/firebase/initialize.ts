/* eslint-disable @typescript-eslint/no-namespace */
import * as admin from 'firebase-admin';
import { AuthConfigService } from '../config/AuthConfigService';

let firebaseApp;
export default (config: AuthConfigService) =>
  firebaseApp
    ? firebaseApp
    : (firebaseApp = admin.initializeApp({
        credential: admin.credential.cert({
          projectId: config.projectId,
          privateKey: config.privateKey.replace(/\\n/g, '\n'),
          clientEmail: config.clientEmail,
        }),
      }));

declare global {
  namespace Express {
    interface Request {
      firebaseUser: admin.auth.DecodedIdToken;
    }
  }
}
