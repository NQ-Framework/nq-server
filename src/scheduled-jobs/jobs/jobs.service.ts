import { Injectable } from '@nestjs/common';
import { SchedulerRegistry } from '@nestjs/schedule';
import { LoggerService } from '../../logger/logger.service';
import { CronJob } from 'cron';
import { ScheduledJob } from "@nqframework/models"
import { HandlerService } from '../handler/handler.service';
import { generateJobId } from "../scheduler/generate-job-id";

@Injectable()
export class JobsService {
    constructor(private logger: LoggerService, private registry: SchedulerRegistry, private handler: HandlerService) {
        this.logger.setContext("Scheduler");
    }

    createJobs(jobsData: ScheduledJob[]) {
        jobsData.forEach(job => {

            const jobId = generateJobId(job);
            const jobDescription = job.organizationId + ": " + job.name;
            const configuration = job.configuration;

            this.logger.log(`creating scheduled job '${jobDescription}' as ${JSON.stringify(job)}`);

            const handler = this.handler.getHandlerFromConfig(configuration);

            if (handler === null) {
                this.logger.error(`Could not load job handler. Job will not be created. requested job type: '${configuration.type}'  details: ${JSON.stringify(job)}`);
                return;
            }

            const invokeHandler = () => {
                handler.ExecuteJob(job.configuration, job);
            }

            const cronJob = new CronJob(job.cronInterval, invokeHandler);

            this.registry.addCronJob(jobId, cronJob);

            cronJob.start();
            this.logger.log(`Started schedueld job: '${jobDescription}'`);
        });
    }

    async deleteExistingJobs(jobsData: ScheduledJob[]) {
        const jobList = (await this.registry.getCronJobs()) as Map<string, CronJob>;
        jobList.forEach((job, id) => {
            const newData = jobsData.find(j => generateJobId(j) === id);
            if (!newData) {
                return;
            }
            this.logger.log(`removing scheduled job '${id}'`);
            job.stop();
            this.registry.deleteCronJob(id);
        });
    }
}
