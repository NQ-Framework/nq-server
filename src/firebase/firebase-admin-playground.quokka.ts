import { loadFirebase } from "./initialize";
import * as dotenv from "dotenv";
dotenv.config({ path: './.development.env' });

const app = loadFirebase({
    projectId: process.env.PROJECT_ID,
    clientEmail: process.env.CLIENT_EMAIL,
    privateKey: process.env.PRIVATE_KEY
} as any);

app.auth().getUserByEmail('milos.s.pfc@gmail.com').then(user => {
    app.auth().setCustomUserClaims(user.uid, {}).then((resp) => {
        console.log('jep', resp);
    });
});
