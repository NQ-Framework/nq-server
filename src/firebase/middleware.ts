import { getFirebaseApp } from './initialize';
import { Injectable, NestMiddleware, HttpStatus } from '@nestjs/common';
import { HttpException } from '@nestjs/common/exceptions/http.exception';
import { Request, Response } from 'express';

@Injectable()
export class FirebaseAuthMiddleware implements NestMiddleware {
  async use(req: Request, _: Response, next: Function) {
    const { authorization } = req.headers;
    if (!authorization) {
      this.throwError();
    }
    const token = authorization.slice(7); // Bearer <token>
    const app = await getFirebaseApp();
    const user = await app
      .auth()
      .verifyIdToken(token)
      .catch((err) => {
        this.throwError(err);
      });

    req.firebaseUser = user;
    next();
  }
  private throwError(err?: Error): never {
    throw new HttpException(
      { message: 'Failed authorization', err },
      HttpStatus.UNAUTHORIZED,
    );
  }
}
