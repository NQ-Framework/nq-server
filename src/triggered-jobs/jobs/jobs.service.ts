import { Injectable } from '@nestjs/common';
import { TriggeredJob } from '@nqframework/models';
import { getFirebaseApp } from "../../firebase/initialize";

@Injectable()
export class JobsService {

    async findJob(jobId: string): Promise<TriggeredJob> {
        const app = await getFirebaseApp();
        const doc = await app.firestore().doc(`triggeredJobs/${jobId}`).get();
        if (!doc) {
            throw new Error('job with given id not found');
        }
        return doc.data() as TriggeredJob;
    }
}
