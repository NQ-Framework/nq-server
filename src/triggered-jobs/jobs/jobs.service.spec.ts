import { Test, TestingModule } from '@nestjs/testing';
import { JobsService } from './jobs.service';
import { getFirebaseApp } from "../../firebase/initialize"
import { TriggeredJob } from '@nqframework/models';
jest.mock("../../firebase/initialize");
const getFirebaseMock = getFirebaseApp as jest.Mock;

const mockJob: TriggeredJob = {
  name: 'triggered job',
  organizationId: 'mock org id',
  id: 'jobId'
} as any
getFirebaseMock.mockImplementation(() => {
  return {
    firestore: () => ({
      doc: (query: string) => {
        return {
          get: () => query === 'triggeredJobs/jobId' ? ({
            data: () => mockJob
          }) : null
        }
      }
    })
  }
});

describe('JobsService', () => {
  let service: JobsService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [JobsService],
    }).compile();

    service = module.get<JobsService>(JobsService);
  });

  it('should return a job by id', async () => {
    const job = await service.findJob('jobId');
    expect(job).toEqual(mockJob)
  });
  it('should throw for non existing job', async () => {
    await expect(service.findJob('invalid jobId')).rejects.toThrowErrorMatchingSnapshot();
  });
});
