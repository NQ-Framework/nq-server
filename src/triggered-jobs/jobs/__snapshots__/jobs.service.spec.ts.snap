// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobsService should throw for non existing job 1`] = `"job with given id not found"`;
